gmp-droid-conv.cpp

```diff
@@ -0,0 +1,237 @@
+
+static void
+CopyPackedPlanes (uint8_t * out0, uint8_t * out1, uint8_t * in, int32_t outSize)
+{
+  int x;
+  uint8_t *place = in;
+  for (x = 0; x < outSize; x++) {
+    out0[x] = place[0];
+    out1[x] = place[1];
+    place += 2;
+  }
+}
+
+class ConvertYUV420PackedSemiPlanar32m:public DroidColourConvert
+{
+public:
+  //TODO: Not tested yet
+  GMPErr Convert (GMPVideoHost * host, DroidMediaData * in,
+      GMPVideoi420Frame * out)
+  {
+    /* copy to the output buffer swapping the u and v planes and cropping if necessary */
+    /* NV12 format with 128 byte alignment */
+
+    uint8_t *y = (uint8_t *) in->data + (m_top * m_stride) + m_left;
+    uint8_t *uv =
+        (uint8_t *) in->data + (m_stride * m_slice_height) +
+        (m_top * m_stride / 2) + m_left / 2;
+
+    GMPPlane *outY, *outU, *outV;
+    // Create plane buffers
+      host->CreatePlane (&outY);
+      host->CreatePlane (&outU);
+      host->CreatePlane (&outV);
+    // Copy Y directly
+      outY->Copy (m_stride * m_height, m_stride, y);
+    // U and V are packed, so we'll have to create empty buffers and copy manually
+      outU->CreateEmptyPlane (m_stride * m_height / 4, m_stride / 2,
+        m_stride * m_height / 4);
+      outV->CreateEmptyPlane (m_stride * m_height / 4, m_stride / 2,
+        m_stride * m_height / 4);
+      CopyPackedPlanes (outU->Buffer (), outV->Buffer (), uv,
+        m_stride * m_height / 4);
+    // Create Frame from the plane buffers to return
+      out->CreateFrame (m_stride * m_height, outY->Buffer (),
+        m_stride * m_height / 4, outU->Buffer (),
+        m_stride * m_height / 4, outV->Buffer (),
+        m_width, m_height, m_stride, m_stride / 2, m_stride / 2);
+    // Destroy the planes
+      outY->Destroy ();
+      outU->Destroy ();
+      outV->Destroy ();
+      return GMPNoErr;
+
+  }
+  void SetFormat (DroidMediaRect * rect, int32_t width, int32_t height)
+  {
+    this->DroidColourConvert::SetFormat (rect, width, height);
+    m_stride = ALIGN_SIZE (m_stride, 128);
+    m_slice_height = ALIGN_SIZE (m_slice_height, 32);
+    m_top = ALIGN_SIZE (m_top, 2);
+    m_left = ALIGN_SIZE (m_left, 2);
+  }
+};
+
+class ConvertYUV420Planar:public DroidColourConvert
+{
+public:
+  GMPErr Convert (GMPVideoHost * host, DroidMediaData * in,
+      GMPVideoi420Frame * out)
+  {
+    /* Buffer is already I420, so we can copy it straight over */
+    /* though we need to handle the cropping using stride and an offset */
+
+    uint8_t *y = (uint8_t *) in->data + (m_top * m_stride) + m_left;
+    uint8_t *u = (uint8_t *) in->data + (m_stride * m_slice_height) +
+        (m_top * m_stride / 2) + (m_left / 2);
+    uint8_t *v = (uint8_t *) in->data + (m_stride * m_slice_height) +
+        (m_stride * m_slice_height / 4)
+        + (m_top * m_stride / 2) + (m_left / 2);
+    // Create plane buffers
+    GMPPlane *outY, *outU, *outV;
+      host->CreatePlane (&outY);
+      host->CreatePlane (&outU);
+      host->CreatePlane (&outV);
```

Comment:
Is the indentation correct?

Reply:
I ran gnu-indent on it. I wonder what confused it: 
indent --braces-on-if-line --case-brace-indentation0 --case-indentation2 --braces-after-struct-decl-line --line-length80 --no-tabs --cuddle-else --dont-line-up-parentheses --continuation-indentation4 --honour-newlines --tab-size8 --indent-level2 --leave-preprocessor-space *.h *.cpp

Reply:
Fixed